docroot/modules/custom/va_gov_user/src/EventSubscriber/UserImport.php

```
@@ -0,0 +1,320 @@
+   *   Environment Discovery service.
+   * @param \Drupal\externalauth\ExternalAuth $externalAuth
+   *   External Authentication service.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   The messenger service.
+   * @param \Drupal\Core\Extension\ModuleHandler $moduleHandler
+   *   Module Handler service.
+   * @param \Drupal\Core\StringTranslation\TranslationInterface $string_translation
+   *   The string translation service.
+   * @param \Drupal\workbench_access\UserSectionStorageInterface $user_section_storage
+   *   The user section storage service.
+   */
+  public function __construct(
+    EntityTypeManager $entity_type_manager,
+    EnvironmentDiscovery $environmentDiscovery,
+    ExternalAuth $externalAuth,
+    MessengerInterface $messenger,
+    ModuleHandler $moduleHandler,
+    TranslationInterface $string_translation,
+    UserSectionStorageInterface $user_section_storage
+  ) {
+    $this->entityTypeManager = $entity_type_manager;
+    $this->environmentDiscovery = $environmentDiscovery;
+    $this->externalAuth = $externalAuth;
+    $this->messenger = $messenger;
+    $this->moduleHandler = $moduleHandler;
+    $this->string_translation = $string_translation;
+    $this->userSectionStorage = $user_section_storage;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    $events[MigrateEvents::POST_ROW_SAVE] = 'onMigratePostRowSave';
+    $events[MigrateEvents::PRE_ROW_SAVE] = 'onMigratePreRowSave';
+    $events[MigrateEvents::POST_IMPORT] = 'onMigratePostImport';
+    return $events;
+  }
+
+  /**
+   * Perform Post Import events.
+   *
+   * @param \Drupal\migrate\Event\MigrateImportEvent $event
+   *   Information about the event that triggered this function.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   * @throws \Drupal\migrate\MigrateException
```

I'm interested in why you're declaring these.  Genuinely interested, mind you, not that arms-crossed-across-my-chest-looking-down-my-nose-at-you-highly-skeptical-and-resistant-to-persuasion way that I so often have re: your pull requests.  I feel like I might really learn something here, so I don't wanna just wander past this.

As I see it, the main value of a @throws declaration is saying what exceptions might be thrown in this function so that some hapless schmuck calling the function will know, like, "okay, this can throw a FileNotFoundException because we're dealing with a calculated file path, but this is 1) Nate's backups so we want to rethrow that 2) Neil's backups so we want to fail silently 3) an SMB wrapper so we want to rethrow that as a different error" or something like that.  Except I don't really do it because I like to keep successive devs on their toes.

But in this case, since we're using an event subscriber, we're basically just documenting this into the void, right?

Or is this a matter of completeness?

Or if we want to test it by simulating some real-world nightmares?

---

wow, I somehow never saw this. Honest answer - it was a copy-paste job and I never questioned the embodied assumptions! Cleaned up in #4358